admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,444 @@
+  {
+    desired_joint_state = reference_joint_state;
+    return controller_interface::return_type::ERROR;
+  }
+
+  // update joint desired joint state
+  for (size_t i = 0; i < reference_joint_state.positions.size(); ++i)
+  {
+    desired_joint_state.positions[i] =
+      reference_joint_state.positions[i] + admittance_state_.joint_pos[i];
+  }
+  for (size_t i = 0; i < reference_joint_state.velocities.size(); ++i)
+  {
+    desired_joint_state.velocities[i] = admittance_state_.joint_vel[i];
+  }
+  for (size_t i = 0; i < reference_joint_state.accelerations.size(); ++i)
+  {
+    desired_joint_state.accelerations[i] = admittance_state_.joint_acc[i];
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+bool AdmittanceRule::calculate_admittance_rule(AdmittanceState & admittance_state, double dt)
+{
+  // create stiffness matrix
+  auto rot_base_control = admittance_state.rot_base_control;
+  Eigen::Matrix<double, 6, 6> K = Eigen::Matrix<double, 6, 6>::Zero();
+  Eigen::Matrix<double, 3, 3> K_pos = Eigen::Matrix<double, 3, 3>::Zero();
+  Eigen::Matrix<double, 3, 3> K_rot = Eigen::Matrix<double, 3, 3>::Zero();
+  K_pos.diagonal() = admittance_state.stiffness.block<3, 1>(0, 0);
+  K_rot.diagonal() = admittance_state.stiffness.block<3, 1>(3, 0);
+  K_pos = rot_base_control * K_pos * rot_base_control.transpose();
+  K_rot = rot_base_control * K_rot * rot_base_control.transpose();
+  K.block<3, 3>(0, 0) = K_pos;
+  K.block<3, 3>(3, 3) = K_rot;
+
+  // create damping matrix
+  Eigen::Matrix<double, 6, 6> D = Eigen::Matrix<double, 6, 6>::Zero();
+  Eigen::Matrix<double, 3, 3> D_pos = Eigen::Matrix<double, 3, 3>::Zero();
+  Eigen::Matrix<double, 3, 3> D_rot = Eigen::Matrix<double, 3, 3>::Zero();
+  D_pos.diagonal() = admittance_state.damping.block<3, 1>(0, 0);
+  D_rot.diagonal() = admittance_state.damping.block<3, 1>(3, 0);
+  D_pos = rot_base_control * D_pos * rot_base_control.transpose();
+  D_rot = rot_base_control * D_rot * rot_base_control.transpose();
+  D.block<3, 3>(0, 0) = D_pos;
+  D.block<3, 3>(3, 3) = D_rot;
+
+  // calculate admittance relative offset
+  Eigen::Isometry3d desired_trans_base_ft;
+  kinematics_->calculate_link_transform(
+    admittance_state.current_joint_pos, admittance_state.ft_sensor_frame, desired_trans_base_ft);
+  Eigen::Matrix<double, 6, 1> X;
+  X.block<3, 1>(0, 0) =
+    desired_trans_base_ft.translation() - admittance_state.ref_trans_base_ft.translation();
+  auto R_ref = admittance_state.ref_trans_base_ft.rotation();
+  auto R_desired = desired_trans_base_ft.rotation();
+  auto R = R_desired * R_ref.transpose();
+  auto angle_axis = Eigen::AngleAxisd(R);
+  X.block<3, 1>(3, 0) = angle_axis.angle() * angle_axis.axis();
+
+  // get admittance relative velocity
+  auto X_dot = Eigen::Matrix<double, 6, 1>(admittance_state.admittance_velocity.data());
+
+  // get external force
+  auto F_base = Eigen::Matrix<double, 6, 1>(admittance_state.wrench_base.data());
+
+  // zero out any forces in the control frame
```

Comment:
What is this block actually doing?

Reply:
`F_base` is just a local variable to shorten `admittance_state.wrench_base` it does nothing else.